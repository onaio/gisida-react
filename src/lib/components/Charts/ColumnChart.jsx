import React from 'react';
import PropTypes from 'prop-types';
import Highcharts from 'highcharts';
import { isNewSeriesData } from './../../utils';

const isNumber = (x) => !Number.isNaN(Number(x));

class ColumnChart extends React.Component {
  static pointFormatterFunc() {
    return `<span>${this.y}%</span>`;
  }

  constructor(props) {
    super(props);
    const {
      seriesTitle,
      seriesData,
      chartHeight,
      chartWidth,
      categories,
      xAxis,
      yAxis,
      yAxisLabel,
      pointFormatterFunc,
      chartType,
      doubleChart,
      chartSpacing,
      chartMargin,
      legendOptions,
      series,
      tooltip,
    } = this.props;

    const { spacingTop, spacingRight, spacingBottom, spacingLeft } = (chartSpacing || {});
    const { marginTop, marginRight, marginBottom, marginLeft } = (chartMargin || {});

    this.state = {
      chart: {
        type: 'column',
        height: chartHeight || null,
        width: chartWidth || null,
        backgroundColor: 'rgba(255,255,255,0)',
        alignTicks: false,
        left:0,
        marginTop: isNumber(marginTop) ? marginTop : 1,
        marginRight: isNumber(marginRight) ? marginRight : 2,
        marginBottom: isNumber(marginBottom) ? marginBottom : 20,
        marginLeft: isNumber(marginLeft) ? marginLeft : 1,
        spacingTop: isNumber(spacingTop) ? spacingTop : 10,
        spacingRight: isNumber(spacingRight) ? spacingRight : 10,
        spacingBottom: isNumber(spacingBottom) ? spacingBottom : 8,
        spacingLeft: isNumber(spacingLeft) ? spacingLeft : 10,
        borderWidth: 0,
        borderRadius:0
      },
      xAxis: (doubleChart === 'multibar' || chartType === 'multi') ? { 
        lineWidth: 0,
        minorGridLineWidth: 0,
        lineColor: 'transparent',
        categories,  
        labels: {
            enabled: true
        },
        minorTickLength: 0,
        tickLength: 0
       } :
        typeof xAxis !== 'undefined'
          ? xAxis || null
          : {
              categories,
              labels: {
                style: {
                  fontSize: 9,
                },
              },
            },
      yAxis: (doubleChart === 'multibar' || chartType === 'multi') ? [{gridLineWidth: 0, visible: false}, ] :
        typeof yAxis !== 'undefined'
          ? yAxis || null
          : [
              {
                title: {
                  text: (yAxisLabel && yAxisLabel) || 'Target Percentage (%)',
                  y: 10,
                },
                endOnTick: false,
              },
              /** The code shows the second label which in some cases maybe undesiarable */
              /**{
                linkedTo: 0,
                opposite: true,
                title: {
                  text: (yAxisLabel && yAxisLabel) || 'Target Percentage (%)',
                  y: 10,
                  x: -10,
               },
            }, **/
            ],
      tooltip: tooltip || {
        useHTML: true,
        shared: true,
        headerFormat: '<b>Range: </b> {point.key} <br/>',
        pointFormat: chartType ? '<tr><td style="color:{series.color};padding:0">{series.name}: </td>' +
        '<td style="padding:0"><b>{point.y:.1f}</b><br/></td></tr>' : null,
        pointFormatter: chartType ? null : pointFormatterFunc ||
          function pointFormatterFunc() {
            return  chartType ? `<tr><td style="color:${this.color};padding:0">${this.name}: </td>' +
            '<td style="padding:0"><b>${this.y}</b></td></tr>` : `<span>${this.y.toLocaleString()}</span>`;
          },
        //shadow: false,
        //backgroundColor: 'transparent',
        //borderWidth: 0,
        //padding: 0,
      },
      title: {
        text: seriesTitle || null,
      },
      credits: {
        enabled: false,
      },
      legend: legendOptions || {
        enabled: true,
      },
      plotOptions: {
        column: {
          showInLegend: legendOptions && typeof legendOptions.enabled !== 'undefined' ? legendOptions.enabled : true,
          pointPadding: doubleChart || chartType === 'multi' ? 0 : 0.2,
          groupPadding: doubleChart || chartType === 'multi' ? 0.1 : 0,
          borderWidth: 0,
          /*tooltip: {
            distance: 0,
            padding: 0,
            pointFormatter: pointFormatterFunc || function pointFormatterFunc() {
              if (isPercent) {
                return `<span>${this.y}%</span>`;
              } else {
                return `<span>${this.y}</span>`;
              }
            },
          },*/
        },
      },
      series: series || (chartType === 'multi' ? seriesData : 
      [
        {	
          name: seriesTitle,	
          data: seriesData,	
        },	
      ]),
    };
  }

  componentDidMount() {
    const self = this;
    
    setTimeout(() => {
      self.chart = Highcharts.chart(self.chartEl, self.state);
    }, 300);
  }

  componentWillReceiveProps(nextProps) {
    const { seriesTitle, seriesData, series, chartHeight, chartWidth, categories, xAxis, chartType, doubleChart } = nextProps;
    
    if (isNewSeriesData(this.state.series[0].data, (seriesData || series[0].data))) {
      if (this.chart) {
        this.chart.destroy();
      }

      this.setState(
        {
          chart: Object.assign({}, this.state.chart, {
            height: chartHeight || null,
            width: chartWidth || null,
          }),
          title: {
            text: seriesTitle || null,
          },
          xAxis: (chartType === 'multi' || doubleChart === 'multibar') ? { 
            lineWidth: 0,
            minorGridLineWidth: 0,
            lineColor: 'transparent',
            categories,    
            labels: {
                enabled: true
            },
            minorTickLength: 0,
            tickLength: 0
           } : typeof xAxis !== 'undefined' ? xAxis || null : {
             categories
           },
            // typeof xAxis !== 'undefined'
            //   ? xAxis || null
            //   : {
            //       categories,
            //       labels: {
            //         style: {
            //           fontSize: 9,
            //         },
            //       },
            //     },
          series: series || chartType === 'multi' ? seriesData : 
          [
            {	
              name: seriesTitle,	
              data: seriesData,	
            },	
          ],
        },
        () => {
          this.chart = Highcharts.chart(this.chartEl, this.state);
        }
      );
    } else if (this.chart && (chartWidth !== this.chart.chartWidth || this.chart.chartHeight)) {
      this.chart.setSize(chartWidth, chartHeight);
    }
  }

  componentWillUnmount() {
    if (this.chart) this.chart.destroy();
  }

  render() {
    return (
      <div
        ref={el => {
          this.chartEl = el;
        }}
      />
    );
  }
}

ColumnChart.propTypes = {
  seriesData: PropTypes.arrayOf(PropTypes.any).isRequired,
  seriesTitle: PropTypes.string.isRequired,
  chartWidth: PropTypes.number.isRequired,
  chartHeight: PropTypes.number.isRequired,
  categories: PropTypes.arrayOf(PropTypes.string).isRequired,
  yAxisLabel: PropTypes.string.isRequired,
};

export default ColumnChart;
