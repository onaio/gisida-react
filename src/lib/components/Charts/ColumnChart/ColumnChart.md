
# ColumnChart
ColumnChart Component is responsible for rendering a column type chart. Based on highcharts, the component takes in props which overide or add chart options. These props control styling and chart views by tweaking the [highcharts api's](https://api.highcharts.com/highcharts/). 
## ColumnChart Props
### Example
```
const props = {
  "categories": [
    "Bay",
    "Nugaal"
  ],
  "isPercent": false,
  "seriesData": [
    {
      "name": "Bay",
      "y": 326,
      "color": "rgba(0, 90, 50, 0.8)"
    },
    {
      "name": "Nugaal",
      "y": 147,
      "color": "rgba(35, 139, 69, 0.8)"
    },
  ],
  "targetMark": 1,
  "chartWidth": 1115,
  "chartHeight": 145,
  "yAxisLabel": "Children treated",
  "showLegend": false
}
<ColumnChart {...props} />
```
The above implementation will render this chart
![ColumnChart Image](https://raw.githubusercontent.com/onaio/gisida-react/staging/public/assets/Screenshot%20from%202020-06-30%2014-22-40.png)

### Props
ColumnChart component takes these props:

**Categories** - Represents the x values on the chart. If not given they will be automatically calculated either starting at 0 and incremented by 1, or from pointStart and pointInterval given in the series options. If you provide the categories, these will be used e.g. ```[
    "Bay",
    "Nugaal"
  ]```

**seriesTitle** - An optional property for secondary charts which renders a title above the chart in the column chart modal

**seriesData** - An array of data points for the series, further explanation on this can be found [here](https://api.highcharts.com/highcharts/series.column.data) e.g ``` [0, 5, 3, 5] ```

**chartHeight, chartWidth, chartSpacing, chartMargin** - General chart options. Extra props can be added to address various styling use cases. More info can be found [here](https://api.highcharts.com/highcharts/chart)

**xAxis** - When using dual or multiple x axes, this number defines which xAxis the particular series is connected to. It refers to either the {@link #xAxis.id|axis id} or the index of the axis in the xAxis array, with 0 being the first. This defaults to 0.

**yAxis** - When using dual or multiple y axes, this number defines which yAxis the particular series is connected to. It refers to either the {@link #yAxis.id|axis id} or the index of the axis in the yAxis array, with 0 being the first. This defaults to 0

**yAxisLabel** - An optional property for column charts to override the default axis title
      
**pointFormatterFunc** - Tooltip formating method. Overrides default highchart tooltipformat e.g 
``` 
pointFormatterFunc() => {
    return `<span>${this.y}%</span>`;
  } 
```

**chartType** - Optional custom prop used to Determines number of charts we are building if 'multi' multiple charts styles are used otherwise picks highcharts defaults 
      
**doubleChart** - Optional Custom prop used to identify if we are building double/ single bar chart. Defaults to single chart. boolean value true/false
      
**legendOptions** - Optional prop that override the symbol creator function and creates custom legend symbols check various options you can tweak on the [legend api](https://api.highcharts.com/highcharts/legend)

**series** - Series options for specific data and the data itself (general options for all series)

**tooltip** - A configuration object for the tooltip rendering of each single series

**showLegend** - optional prop that decides whether or not to show chart legends. Defaults to true.
