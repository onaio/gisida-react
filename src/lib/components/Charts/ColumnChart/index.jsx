import React from 'react';
import PropTypes from 'prop-types';
import Highcharts from 'highcharts';
import { isNewSeriesData } from '../../../utils';

export const isNumber = (x) => !Number.isNaN(Number(x));

class ColumnChart extends React.Component {
  static pointFormatterFunc() {
    return `<span>${this.y}%</span>`;
  }
  constructor(props) {
    super(props);
    const {
      seriesTitle,
      seriesData,
      chartHeight,
      chartWidth,
      categories,
      xAxis,
      yAxis,
      yAxisLabel,
      pointFormatterFunc,
      chartType,
      doubleChart,
      chartSpacing,
      chartMargin,
      legendOptions,
      series,
      tooltip,
      showLegend
    } = this.props;
    const { spacingTop, spacingRight, spacingBottom, spacingLeft } = (chartSpacing || {});
    const { marginTop, marginRight, marginBottom, marginLeft } = (chartMargin || {});
    this.state = {
      chart: {
        type: 'column',
        height: chartHeight || null,
        width: chartWidth || null,
        backgroundColor: 'rgba(255,255,255,0)',
        alignTicks: false,
        left:0,
        marginTop: isNumber(marginTop) ? marginTop : null,
        marginRight: isNumber(marginRight) ? marginRight : null,
        marginBottom: isNumber(marginBottom) ? marginBottom : null,
        marginLeft: isNumber(marginLeft) ? marginLeft : null,
        spacingTop: isNumber(spacingTop) ? spacingTop : 15,
        spacingRight: isNumber(spacingRight) ? spacingRight : 10,
        spacingBottom: isNumber(spacingBottom) ? spacingBottom : null,
        spacingLeft: isNumber(spacingLeft) ? spacingLeft : null,
        borderWidth: 0,
        borderRadius:0
      },
      xAxis: (doubleChart === 'multibar' || chartType === 'multi') ? { 
        lineWidth: 0,
        minorGridLineWidth: 0,
        lineColor: 'transparent',
        categories,  
        labels: {
            enabled: true
        },
        minorTickLength: 0,
        tickLength: 0
       } :
        typeof xAxis !== 'undefined'
          ? xAxis || null
          : {
              categories,
              labels: {
                style: {
                  fontSize: 9,
                },
              },
            },
      yAxis: (doubleChart === 'multibar' || chartType === 'multi') ? [{gridLineWidth: 0, visible: false}, ] :
        typeof yAxis !== 'undefined'
          ? yAxis || null
          : [
              {
                title: {
                  text: (yAxisLabel && yAxisLabel) || 'Target Percentage (%)',
                  y: 10,
                },
                endOnTick: false,
              },
              /** The code shows the second label which in some cases maybe undesiarable */
              {
                linkedTo: 0,
                opposite: true,
                title: {
                  text: (yAxisLabel && yAxisLabel) || 'Target Percentage (%)',
                  y: 10,
                  x: -10,
               },
            },
            ],
      tooltip: tooltip || {
        useHTML: true,
        shared: true,
        headerFormat: doubleChart ?  '<b>Range: </b> {point.key} <br/>' : '<b>{point.key}: </b>',
        pointFormat: chartType ? '<tr><td style="color:{series.color};padding:0">{series.name}: </td>' +
        '<td style="padding:0"><b>{point.y:.1f}</b><br/></td></tr>' : '',
        pointFormatter: chartType ? null : pointFormatterFunc ||
          function pointFormatterFunc() {
            return  chartType ? `<tr><td style="color:${this.color};padding:0">${this.name}: </td>' +
            '<td style="padding:0"><b>${this.y}</b></td></tr>` : `<span>${this.y.toLocaleString()}</span>`;
          },
        shadow: doubleChart ? false : true,
        backgroundColor: doubleChart ? 'transparent' : undefined,
        borderWidth: doubleChart ? 0 : 1,
        padding: doubleChart ? 0 : 8,
      },
      title: {
        text: seriesTitle || null,
      },
      credits: {
        enabled: false,
      },
      legend: legendOptions || {
        enabled: showLegend ? showLegend : true,
      },
      plotOptions: {
        column: {
          showInLegend: legendOptions && typeof legendOptions.enabled !== 'undefined' ? legendOptions.enabled : false,
          pointPadding: doubleChart || chartType === 'multi' ? 0 : 0.2,
          groupPadding: doubleChart || chartType === 'multi' ? 0.1 : 0.2,
          borderWidth: 0
        },
      },
      series: series || (chartType === 'multi' ? seriesData : 
      [
        {	
          name: seriesTitle,	
          data: seriesData,	
        },	
      ]),
    };
  }

  componentDidMount() {
    const self = this;
    
    setTimeout(() => {
      self.chart = Highcharts.chart(self.chartEl, self.state);
    }, 300);
  }

  componentWillReceiveProps(nextProps) {
    const { seriesTitle, seriesData, series, chartHeight, chartWidth, categories, xAxis, chartType, doubleChart } = nextProps;
    
    if (isNewSeriesData(this.state.series[0].data, (seriesData || series[0].data))) {
      if (this.chart) {
        this.chart.destroy();
      }

      this.setState(
        {
          chart: Object.assign({}, this.state.chart, {
            height: chartHeight || null,
            width: chartWidth || null,
          }),
          title: {
            text: seriesTitle || null,
          },
          xAxis: (chartType === 'multi' || doubleChart === 'multibar') ? { 
            lineWidth: 0,
            minorGridLineWidth: 0,
            lineColor: 'transparent',
            categories,    
            labels: {
                enabled: true
            },
            minorTickLength: 0,
            tickLength: 0
           } : typeof xAxis !== 'undefined' ? xAxis || null : {
             categories,
             labels: {
              style: {
                fontSize: 9,
              },
            },
           },
          series: series || chartType === 'multi' ? seriesData : 
          [
            {	
              name: seriesTitle,	
              data: seriesData,	
              animation: {
                duration: 0,
              },
            },	
          ],
        },
        () => {
          this.chart = Highcharts.chart(this.chartEl, this.state);
        }
      );
    } else if (this.chart && (chartWidth !== this.chart.chartWidth || this.chart.chartHeight)) {
      this.chart.setSize(chartWidth, chartHeight);
    }
  }

  componentWillUnmount() {
    if (this.chart) this.chart.destroy();
  }

  render() {
    return (
      <div
        ref={el => {
          this.chartEl = el;
        }}
      />
    );
  }
}

ColumnChart.propTypes = {
  seriesData: PropTypes.arrayOf(PropTypes.any).isRequired,
  seriesTitle: PropTypes.string.isRequired,
  chartWidth: PropTypes.number.isRequired,
  chartHeight: PropTypes.number.isRequired,
  categories: PropTypes.arrayOf(PropTypes.string).isRequired,
  yAxisLabel: PropTypes.string.isRequired,
};
export const pointFormatterFunc = ColumnChart.pointFormatterFunc;
export default ColumnChart;
