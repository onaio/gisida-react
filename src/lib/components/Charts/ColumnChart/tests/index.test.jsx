import React from 'react';
import ColumnChart, { isNumber, pointFormatterFunc } from '..';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('ColumnChart component', () => {
  const props =  {
        "categories": [
          "Bay",
          "Nugaal"
        ],
        "isPercent": false,
        "seriesData": [
          {
            "name": "Bay",
            "y": 326,
            "color": "rgba(0, 90, 50, 0.8)"
          },
          {
            "name": "Nugaal",
            "y": 147,
            "color": "rgba(35, 139, 69, 0.8)"
          }
        ],
        "chartWidth": 1115,
        "chartHeight": 145,
        "yAxisLabel": "Children treated",
        "showLegend": false
      }
  const componentWrapper = shallow(
    <ColumnChart {...props} />
  );

  it('component renderes correctly', () => {
    const json = toJson(componentWrapper)
    expect(json).toMatchSnapshot();
  });
  it('renders and passes props correctly', () => {
    const wrapper = mount(<ColumnChart {...props}/>);
    expect(wrapper.props()).toMatchSnapshot('columnchart props')
  });
  it('checks if it\'s a number', () => {
      expect(isNumber('hello')).toEqual(false);
      expect(isNumber(100)).toEqual(true);
  })
});
