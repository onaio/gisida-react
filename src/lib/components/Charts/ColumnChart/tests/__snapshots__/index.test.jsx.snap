// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnChart component component renderes correctly 1`] = `<div />`;

exports[`ColumnChart component renders and passes props correctly: columnchart props 1`] = `
Object {
  "categories": Array [
    "Bay",
    "Nugaal",
  ],
  "chartHeight": 145,
  "chartWidth": 1115,
  "isPercent": false,
  "seriesData": Array [
    Object {
      "color": "rgba(0, 90, 50, 0.8)",
      "name": "Bay",
      "y": 326,
    },
    Object {
      "color": "rgba(35, 139, 69, 0.8)",
      "name": "Nugaal",
      "y": 147,
    },
  ],
  "showLegend": false,
  "yAxisLabel": "Children treated",
}
`;
